tokio/src/sync/mutex.rs

```diff
@@ -86,11 +6,89 @@ use std::error::Error;
 use std::fmt;
 use std::ops::{Deref, DerefMut};
 
-/// An asynchronous mutual exclusion primitive useful for protecting shared data
+/// An asynchronous `Mutex`-like type.
 ///
-/// Each mutex has a type parameter (`T`) which represents the data that it is protecting. The data
-/// can only be accessed through the RAII guards returned from `lock`, which
-/// guarantees that the data is only ever accessed when the mutex is locked.
+/// This type acts similarly to an asynchronous [`std::sync::Mutex`], with one
+/// major difference: [`lock`] does not block.  
+/// Another difference is [`lock`] can be held across await points.
+/// This allows you to do something along the lines of:
+///
+/// ```rust,no_run
+/// use tokio::sync::Mutex;
+/// use std::sync::Arc;
+///
+/// #[tokio::main]
+/// async fn main() {
+///     let data1 = Arc::new(Mutex::new(0));
+///     let data2 = Arc::clone(&data1);
+///
+///     tokio::spawn(async move {
+///         let mut lock = data2.lock().await;
+///         *lock += 1;
+///     });
+///
+///     let mut lock = data1.lock().await;
+///     *lock += 1;
+/// }
+/// ```
+///
+/// Another example
+/// ```rust,no_run
+/// #![warn(rust_2018_idioms)]
+///
+/// use tokio::sync::Mutex;
+/// use std::sync::Arc;
+///
+///
+/// #[tokio::main]
+/// async fn main() {
+///     let count = Arc::new(Mutex::new(0));
+///
+///     for _ in 0..5 {
+///         let my_count = Arc::clone(&count);
+///         tokio::spawn(async move {
+///             for _ in 0..10 {
+///                 let mut lock = my_count.lock().await;
+///                 *lock += 1;
+///                 println!("{}", lock);
+///             }
+///         });
+///     }
+///
+///     loop {
+///         if *count.lock().await >= 50 {
+///             break;
+///         }
+///     }
+///    println!("Count hit 50.");
+/// }
+/// ```
+/// There are a few things of note here to pay attention to in this example.
+/// 1. The mutex is wrapped in an [`Arc`] to allow it to be shared across threads.
+/// 2. Each spawned task obtains a lock and releases it on every iteration.
+/// 3. Mutation of the data the Mutex is protecting is done by de-referencing the the obtained lock
+///    as seen on lines 23 and 30.
+///
+/// Tokio's Mutex works in a simple FIFO (first in, first out) style where all calls
+/// to lock complete in the order they were performed. In that way
+/// the Mutex is "fair" and predictable in how it distributes the locks to inner data. This is why
+/// the output of this program is an in-order count to 50. Locks are released and reacquired
+/// after every iteration, so basically, each thread goes to the back of the line after it increments
+/// the value once. Also, since there is only a single valid lock at any given time there is no
+/// possibility of a race condition when mutating the inner value.
```

Comment:
```suggestion
/// the value once. Finally, since there is only a single valid lock at any given time, there is no
/// possibility of a race condition when mutating the inner value.
```